### PR TITLE
[MINOR][PYSPARK][DOCS] Fix typo in example documentation

### PR DESCRIPTION
**Pull request by mstill3**
### What changes were proposed in this pull request?

I propose that we change the example code documentation to call the proper function .
For example, under the `foreachBatch` function, the example code was calling the `foreach()` function by mistake.

### Why are the changes needed?

I suppose it could confuse some people, and it is a typo

### Does this PR introduce any user-facing change?

No, there is no "meaningful" code being change, simply the documentation

### How was this patch tested?

I made the change on a fork and it still worked

### CODE REVIEW

**Comment by HyukjinKwon** [comment]
ok to test

**Comment by HyukjinKwon** [comment]
Can you check if there are some more typos while we're here? I am sure there are more.

File: python/pyspark/sql/session.py [hunk 1]
```diff
@@ -94,12 +94,12 @@ def config(self, key=None, value=None, conf=None):
 
             >>> from pyspark.conf import SparkConf
             >>> SparkSession.builder.config(conf=SparkConf())
-            <pyspark.sql.session...
+            <pyspark.sql.session.SparkSession.Builder at 0x7fa346504c50>
```
Comment (srowen) [inline python/pyspark/sql/session.py]: Hm, I don't think you want to try to match this in the doctest? that's what the ... was for, I thought

File: python/pyspark/sql/dataframe.py [hunk 2]
```diff
@@ -1077,7 +1077,7 @@ def sort(self, *cols, **kwargs):
         """Returns a new :class:`DataFrame` sorted by the specified column(s).
 
         :param cols: list of :class:`Column` or column names to sort by.
-        :param ascending: boolean or list of boolean (default True).
+        :param ascending: boolean or list of boolean (default ``True``).
```
Comment (srowen) [inline python/pyspark/sql/dataframe.py]: What does the double-back-tick do, out of curiosity?

**Review by srowen — CHANGES_REQUESTED** [review]
Yep several good catches in there

File: python/pyspark/sql/session.py [hunk 3]
```diff
@@ -624,20 +624,20 @@ def createDataFrame(self, data, schema=None, samplingRatio=None, verifySchema=Tr
...
         :class:`pyspark.sql.types.StructType`, it will be wrapped into a
-        :class:`pyspark.sql.types.StructType` as its only field, and the field name will be "value",
-        each record will also be wrapped into a tuple, which can be converted to row later.
+        :class:`pyspark.sql.types.StructType` as its only field, and the field name will be "value".
+        Each record will also be wrapped into a tuple, which can be converted to row later.
 
         If schema inference is needed, ``samplingRatio`` is used to determined the ratio of
         rows used for schema inference. The first row will be used if ``samplingRatio`` is ``None``.
 
         :param data: an RDD of any kind of SQL data representation(e.g. row, tuple, int, boolean,
```
[the remark this answers is not in the document]
Comment (mstill3) [reply python/pyspark/sql/session.py]: okay can do!

File: python/pyspark/sql/dataframe.py (hunk 2 above)
Comment (mstill3) [reply python/pyspark/sql/dataframe.py]: It formats the True to show up as slightly different text that is for python code expressions

File: python/pyspark/sql/session.py (hunk 1 above)
Comment (mstill3) [reply python/pyspark/sql/session.py]: Good call

File: python/pyspark/sql/readwriter.py [hunk 4]
```diff
@@ -733,7 +733,7 @@ def save(self, path=None, format=None, mode=None, partitionBy=None, **options):
         :param partitionBy: names of partitioning columns
         :param options: all other string options
 
-        >>> df.write.mode('append').parquet(os.path.join(tempfile.mkdtemp(), 'data'))
+        >>> df.select("name", "favorite_color").write.save("namesAndFavColors.parquet")
```
Comment (dongjoon-hyun) [inline python/pyspark/sql/readwriter.py]: This is a code change which should not be mixed into here.
Please revert this change.

File: python/pyspark/sql/session.py [hunk 5]
```diff
@@ -136,7 +136,7 @@ def appName(self, name):
         @since(2.0)
         def enableHiveSupport(self):
             """Enables Hive support, including connectivity to a persistent Hive metastore, support
-            for Hive serdes, and Hive user-defined functions.
+            for Hive SerDe's, and Hive user-defined functions.
```
Comment (dongjoon-hyun) [inline python/pyspark/sql/session.py]: ? This should not be a possessive form. Please revert this change.

File: python/pyspark/sql/session.py [hunk 6]
```diff
@@ -151,7 +151,7 @@ def getOrCreate(self):
             new one based on the options set in this builder.
 
             This method first checks whether there is a valid global default SparkSession, and if
-            yes, return that one. If no valid global default SparkSession exists, the method
+            there is, return that one. If no valid global default SparkSession exists, the method
```
Comment (dongjoon-hyun) [inline python/pyspark/sql/session.py]: If you change this from `yes` to `there is`, it breaks the remaining `If no`.
I'd like to recommend not to do this in this PR.

File: python/pyspark/sql/readwriter.py (hunk 4 above)
Comment (srowen) [reply python/pyspark/sql/readwriter.py]: I think it's a fix - this is doc for `.save` and this example doesn't match

File: python/pyspark/sql/streaming.py [hunk 7]
```diff
@@ -1062,7 +1062,7 @@ def foreachBatch(self, func):
         >>> def func(batch_df, batch_id):
         ...     batch_df.collect()
         ...
-        >>> writer = sdf.writeStream.foreach(func)
+        >>> writer = sdf.writeStream.foreachBatch(func)
```
Comment (dongjoon-hyun) [inline python/pyspark/sql/streaming.py]: Nice catch!

File: python/pyspark/sql/readwriter.py (hunk 4 above)
Comment (dongjoon-hyun) [reply python/pyspark/sql/readwriter.py]: Oh. Got it. Thanks!

**Comment by dongjoon-hyun** [comment]
Thank you for your first contribution, @mstill3 .  I left a few minor comments.
BTW, please don't change the title. There is a rule for that. :)

We use `[SPARK-JIRAID][COMPONENT-NAME]` style. Since you don't make a JIRA issue, we use `[MINOR]` instead of `[SPARK-JIRAID]`.

File: python/pyspark/sql/streaming.py (hunk 7 above)
Comment (mstill3) [reply python/pyspark/sql/streaming.py]: Thanks!

File: python/pyspark/sql/session.py (hunk 5 above)
Comment (srowen) [reply python/pyspark/sql/session.py]: SerDes is fine though

File: python/pyspark/sql/session.py (hunk 6 above)
Comment (srowen) [reply python/pyspark/sql/session.py]: How about just "If not," to avoid repeating the predicate.

**Comment by dongjoon-hyun** [comment]
Could you run `dev/lint-python` and fix the errors?
```
pycodestyle checks failed:
80
./python/pyspark/sql/session.py:154:97: W291 trailing whitespace
81
./python/pyspark/sql/session.py:155:98: W291 trailing whitespace
82
##[error]Process completed with exit code 1.
```

File: python/pyspark/sql/dataframe.py [hunk 8]
```diff
@@ -1975,7 +1980,7 @@ def withColumn(self, colName, col):
         existing column that has the same name.
 
         The column expression must be an expression over this DataFrame; attempting to add
-        a column from some other dataframe will raise an error.
+        a column from some other DataFrame will raise an error.
```
Comment (dongjoon-hyun) [inline python/pyspark/sql/dataframe.py]: Maybe, did you want the following?
```
a column from some other :class:`DataFrame` will raise an error.
```

File: python/pyspark/sql/session.py [hunk 9]
```diff
@@ -151,9 +151,9 @@ def getOrCreate(self):
             new one based on the options set in this builder.
 
             This method first checks whether there is a valid global default SparkSession, and if
-            yes, return that one. If no valid global default SparkSession exists, the method
-            creates a new SparkSession and assigns the newly created SparkSession as the global
-            default.
+            there is, return that one. Otherwise if no valid global default SparkSession exists,
```
Comment (dongjoon-hyun) [inline python/pyspark/sql/session.py]: `Otherwise if`? Shall we keep the original sentences for this part?

File: python/pyspark/sql/dataframe.py (hunk 8 above)
Comment (mstill3) [reply python/pyspark/sql/dataframe.py]: Yeah good catch

File: python/pyspark/sql/readwriter.py (hunk 4 above)
Comment (dongjoon-hyun) [inline python/pyspark/sql/readwriter.py]: @mstill3 . As you see, this causes UT failure because `df` has only `[day, month, name, year]`.
```
Failed example:
    df.select("name", "favorite_color").write.save("namesAndFavColors.parquet")
Exception raised:
    Traceback (most recent call last):
      File "/usr/lib64/pypy-2.5.1/lib-python/2.7/doctest.py", line 1315, in __run
        compileflags, 1) in test.globs
      File "<doctest pyspark.sql.readwriter.DataFrameWriter.save[0]>", line 1, in <module>
        df.select("name", "favorite_color").write.save("namesAndFavColors.parquet")
      File "/home/jenkins/workspace/SparkPullRequestBuilder@2/python/pyspark/sql/dataframe.py", line 1325, in select
        jdf = self._jdf.select(self._jcols(*cols))
      File "/home/jenkins/workspace/SparkPullRequestBuilder@2/python/lib/py4j-0.10.8.1-src.zip/py4j/java_gateway.py", line 1286, in __call__
        answer, self.gateway_client, self.target_id, self.name)
      File "/home/jenkins/workspace/SparkPullRequestBuilder@2/python/pyspark/sql/utils.py", line 102, in deco
        raise converted
    AnalysisException: cannot resolve '`favorite_color`' given input columns: [day, month, name, year];;
```

**Review by dongjoon-hyun — CHANGES_REQUESTED** [review]
The UT failure seems to be the last blocker for this PR. Could you fix that in this PR?

File: python/pyspark/sql/readwriter.py [hunk 10]
```diff
@@ -733,7 +733,7 @@ def save(self, path=None, format=None, mode=None, partitionBy=None, **options):
         :param partitionBy: names of partitioning columns
         :param options: all other string options
 
-        >>> df.write.mode('append').parquet(os.path.join(tempfile.mkdtemp(), 'data'))
+        >>> df.select("day", "month", "name", "year").write.save("mydata.parquet")
```
Comment (dongjoon-hyun) [inline python/pyspark/sql/readwriter.py]: I guess we need `mode("overwrite")`.
```
File "/home/jenkins/workspace/SparkPullRequestBuilder/python/pyspark/sql/readwriter.py", line 735, in pyspark.sql.readwriter.DataFrameWriter.save
Failed example:
    df.select("day", "month", "name", "year").write.save("mydata.parquet")
Exception raised:
    Traceback (most recent call last):
      File "/home/anaconda/lib/python2.7/doctest.py", line 1315, in __run
        compileflags, 1) in test.globs
      File "<doctest pyspark.sql.readwriter.DataFrameWriter.save[0]>", line 1, in <module>
        df.select("day", "month", "name", "year").write.save("mydata.parquet")
      File "/home/jenkins/workspace/SparkPullRequestBuilder/python/pyspark/sql/readwriter.py", line 746, in save
        self._jwrite.save(path)
      File "/home/jenkins/workspace/SparkPullRequestBuilder/python/lib/py4j-0.10.8.1-src.zip/py4j/java_gateway.py", line 1286, in __call__
        answer, self.gateway_client, self.target_id, self.name)
      File "/home/jenkins/workspace/SparkPullRequestBuilder/python/pyspark/sql/utils.py", line 102, in deco
        raise converted
    AnalysisException: path file:/home/jenkins/workspace/SparkPullRequestBuilder/mydata.parquet already exists.;
```

Comment (dongjoon-hyun) [reply python/pyspark/sql/readwriter.py]: BTW, let's remove `.select("day", "month", "name", "year")`. We don't need to `select` because this is `save` API doc.

File: python/pyspark/sql/readwriter.py [hunk 11]
```diff
@@ -733,7 +733,7 @@ def save(self, path=None, format=None, mode=None, partitionBy=None, **options):
         :param partitionBy: names of partitioning columns
         :param options: all other string options
 
-        >>> df.write.mode('append').parquet(os.path.join(tempfile.mkdtemp(), 'data'))
+        >>> df.write.mode("overwrite").save("mydata.parquet")
```
Comment (HyukjinKwon) [inline python/pyspark/sql/readwriter.py]: I think we should revert this change. Tmp directory will be created under target directory as of https://github.com/apache/spark/commit/a634d66ce767bd5e1d8553d1a2c32e2b1a80f642 ; however, now seems it will create under the current working directory.

**Review by HyukjinKwon — APPROVED** [review]
LGTM otherwise.

File: python/pyspark/sql/readwriter.py (hunk 11 above)
Comment (srowen) [reply python/pyspark/sql/readwriter.py]: The change is really to use `.save`, as that what this is supposed to be an example of. If it makes sense to make a temp file to save to, yes I agree, leave that part the same.

Comment (HyukjinKwon) [reply python/pyspark/sql/readwriter.py]: Oh, yes. We should use `.save`. Thanks for clarification.

**Comment by dongjoon-hyun** [comment]
@mstill3 . Please update the PR to use `os.path.join(tempfile.mkdtemp(), 'data'` according to @HyukjinKwon 's advice.

**Review by dongjoon-hyun — APPROVED** [review]
LGTM for the others.

**Review by srowen — APPROVED** [review]

**Review by dongjoon-hyun — APPROVED** [review]
Thank you again for your first contribution, @mstill3 .
Thank you, @srowen, @HyukjinKwon , and @vegerot
Merged to master.

**Comment by dongjoon-hyun** [comment]
@mstill3 . Since this is useful for users, could you make a PR for `branch-2.4` too?
